File: C++/Graphs/27_RedundantConnection.cpp
int find(int x, vector<int> &parent)
{
    if (parent[x] == -1)
        return x;
    return parent[x] = find(parent[x], parent);
}
void unionSet(int x, int y, vector<int> &parent, vector<int> &rank)
{
    int h1 = find(x, parent);
    int h2 = find(y, parent);
    if (h1 != h2)
    {
        if (rank[h1] < rank[h2])
        {
            parent[h1] = h2;
            rank[h2] += rank[h1];
        }
        else
        {
            parent[h2] = h1;
            rank[h1] += rank[h2];
        }
    }
}

vector<int> findRedundantConnection(vector<vector<int>> &edges)
{
    vector<int> ans;

    int n = edges.size();
    vector<int> parent(n + 1, -1);
    vector<int> rank(n + 1, 1);
    for (auto edge : edges)
    {
        int x = find(edge[0], parent);
        int y = find(edge[1], parent);

        if (x != y)
        {
            unionSet(x, y, parent, rank);
        }
        else
        {
            ans.push_back(edge[0]);
            ans.push_back(edge[1]);
        }
    }
    int c = ans.size();
    return {ans[c - 2], ans[c - 1]};
}

File: Java/Graphs/27_RedundantConnection.java
class Solution {
    int par[];
    public int findPar(int u){
        if(u == par[u]) return u;
        return par[u] = findPar(par[u]);
    }
    public int[] findRedundantConnection(int[][] edges) {
        int n = edges.length;
        int ans[] = new int[2];
        par = new int[n+1];
        for(int i=1;i<=n;i++){
            par[i] = i;
        }
        
        for(int[] a: edges){
            int gp1 = findPar(a[0]);
            int gp2 = findPar(a[1]);
            if(gp1 != gp2){
                par[gp2] = gp1;
            }
            else{
                ans[0] = a[0];
                ans[1] = a[1];
                break;
            }
        }
        return ans;
    }
}